src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -469,6 +472,16 @@ public ProjectState UpdateParseOptions(ParseOptions options)
                 documentStates: docMap);
         }
 
+        public ProjectState UpdateCompilationOutputs(CompilationOutputs outputs)
```

Comment:
So this is a *bit* funky -- this doesn't quite follow the usual snapshot model in that CompilationOutputs's stream methods actually return that doesn't match up with the current snapshot but something older -- it's implicitly the result _of the last time we built_, right? It also means if somebody does observe the value, we won't necessarily be persisting it for later calls to ensure immutability.

I can think of a few options:

1. Don't care, and just do nothing.
2. Instead make an ICompilationOutputsWorkspaceService (which would still be public) and that just has a GetCompilationOutputs method that takes a ProjectId. That way it's no longer strictly tied to the snapshot and doesn't come with the guarantees thereof.

I'm thinking 2 would be my preference. The CompilationOutputs type you've got itself I think would stay unchanged and all the work already done up in VisualStudioProject.

Reply:
How is this different from having OutputPath on Project? The CompilationOuputs instance is a an abstracted set of "paths".

Reply:
Thinking about this more I agree with going with [2].